### DISCUSSION TITLE
Custom integration issue - Design (css) set in project settings is not applied

### DISCUSSION THREAD

**Discussion by sukapyx** [mosparo/mosparo#235] [Q&A]
Hello. Im running Mosparo on localhost using IIS. I have created a simple html page where i use the Mosparo box. Everything seems fine, the connection to API also works, no errors on the console. Only problem is that for some reason the Design (specified in project settings) does not apply.

My IIS setup:
![image](https://github.com/mosparo/mosparo/assets/143611950/3b2fd1f9-8649-4915-9490-dee9dbe5f6bb)
In addition I have granted IIS access to Windows/Temp folder

My website and Mosparo installation (public folder) where I also have my custom index.html and style.css. I needed to create the site outside of IIS standard inetpub folder, since IIS access right prbolems which i was unable to solve.:
![image](https://github.com/mosparo/mosparo/assets/143611950/263acdef-fc15-4b89-ba8b-8ff886b31849)

Index.html content:
![image](https://github.com/mosparo/mosparo/assets/143611950/2c17dd2a-d427-4a5b-9c5b-1a5c29302302)

![image](https://github.com/mosparo/mosparo/assets/143611950/9f491f5c-3f80-4635-aad8-ee8b14456bd6)

Project settings Design configuration
![image](https://github.com/mosparo/mosparo/assets/143611950/31c57ab2-6cbb-45eb-84f9-571a9f8f527e)

API works:
![image](https://github.com/mosparo/mosparo/assets/143611950/02549705-66f2-4f1b-b8fc-8b76af9af569)

MySQL database table content:
![image](https://github.com/mosparo/mosparo/assets/143611950/74a5427a-7a91-4c18-9c35-2ddf414fd990)

resources folder:
![image](https://github.com/mosparo/mosparo/assets/143611950/2b64b426-e798-44f7-8f0b-a8783a4a728d)

this is empty for some reason - is it supposed to be?
![image](https://github.com/mosparo/mosparo/assets/143611950/b1260ce1-ee47-4d52-be25-47ca22ef4cf5)

Im running out of ideas here :|

**Comment by zepich** [comment]
Hi @sukapyx 

Thank you very much for your question.

Your last screenshot is the most important one. No, the directory should not be empty.

If the directory is empty, mosparo will not redirect to the correct CSS file and will fall back to the default one.

The question is: why is the file not saved correctly?

Can you try to save the design settings once and then check the log file in `var/log/prod-{date}.log`? Is there any error visible?

One additional thing: If you're using `loadCssResource` in the mosparo function, you do not have to include the mosparo CSS file yourself (as visible in the 3rd screenshot).

Please let me know if you see any errors in the log file.

Kind regards,

zepich

**Reply by sukapyx** [reply]
Thank you for your reply @zepich 

I modified the project Design settings and saved the changes, this produces no errors in the log file you mentioned. The DB records are updated, also the mappings.php file:
![image](https://github.com/mosparo/mosparo/assets/143611950/472c3c6a-e52b-40e3-8efa-a336043ac1c9)
![image](https://github.com/mosparo/mosparo/assets/143611950/bd9b5539-ccbd-42c0-b43f-744708792c25)

Also I removed the inclusion as you advised.
![image](https://github.com/mosparo/mosparo/assets/143611950/8b48c372-915b-44fb-a83a-9ff6311d3fed)

Not sure if this is relevant, but I'm running into problems also when updating (all reguirements, but the two optional plugins, are met) but it halts the update process (produces 500 internal server error on the console).
![image](https://github.com/mosparo/mosparo/assets/143611950/fe16d86b-6021-41d6-bd85-b711c413e39d)

[2024-05-30T11:38:47.463982+00:00] php.CRITICAL: Fatal Error: Maximum execution time of 30 seconds exceeded {"exception":"[object] (Symfony\\Component\\ErrorHandler\\Error\\FatalError(code: 0): Error: Maximum execution time of 30 seconds exceeded at C:\\MyTestWebsite\\src\\Helper\\UpdateHelper.php:803)"} []
[2024-05-30T11:38:47.469124+00:00] request.CRITICAL: Uncaught PHP Exception Symfony\Component\ErrorHandler\Error\FatalError: "Error: Maximum execution time of 30 seconds exceeded" at C:\MyTestWebsite\src\Helper\UpdateHelper.php line 803 {"exception":"[object] (Symfony\\Component\\ErrorHandler\\Error\\FatalError(code: 0): Error: Maximum execution time of 30 seconds exceeded at C:\\MyTestWebsite\\src\\Helper\\UpdateHelper.php:803)"} []


Also including the General project settings, as im not sure if the hosts part is actually correct:
![image](https://github.com/mosparo/mosparo/assets/143611950/7f62fb6e-a5bb-4797-8bc8-3febe4c5ab25)


Best regards

**Reply by zepich** [reply]
Hi @sukapyx 

Thank you very much for your feedback.

Let's talk about the update issue after fixing the issue with the CSS file.

Can you adjust the following file?

File: `src/Helper/DesignHelper.php`
Line: 546 (in v1.0) / 548 (in v1.1) / 561 (in v1.2)

```php
                 $this->filesystem->remove($file->getPathname());
```

Can you comment this line (add `//` at the beginning of the line) and then try to save the design settings again?

Kind regards,

zepich

**Reply by zepich** [reply]
(Regarding the hosts: Let's focus on the CSS file issue first and then discuss the other issues to not mix up the discussion)

**Reply by sukapyx** [reply]
After commenting that line out: 
1. no changes in the log file
2. change in the UI and error on the console
![image](https://github.com/mosparo/mosparo/assets/143611950/59138925-c150-4c99-b9be-6c30fa1adb04)
3. the missing .css file has appeared under to the resources project guid folder
![image](https://github.com/mosparo/mosparo/assets/143611950/2e1eb0ec-a955-4cb4-b33c-abcaae22869b)

![image](https://github.com/mosparo/mosparo/assets/143611950/93e5db9f-54f7-4614-b2b6-90c83987f3ca)

**Reply by zepich** [reply]
Hi @sukapyx 

Okay, perfect, great news! Thank you very much.

I will test the DesignHelper issue and provide a correct fix with the next mosparo version.

Now, the next issue: Redirecting. Another user had the same problem. This user posted the `web.config` file regarding the URL rewriting. Can you check if you have similar settings or show me your `web.config` file?

```xml
<?xml version="1.0" encoding="UTF-8"?>
<configuration>
	<system.webServer>
		<cors enabled="true" failUnlistedOrigins="true">
			<add origin="*" />
		</cors>
		<handlers>
			<add name="php8.1 phalcon5" path="*.php" verb="*" modules="FastCgiModule" scriptProcessor="C:\server\php\php-8.1.27-nts-Win32-vs16-x64\php-cgi.exe" resourceType="File" requireAccess="Script" />
		</handlers>
		<defaultDocument>
			<files>
				<add value="index.php" />
			</files>
		</defaultDocument>
		<rewrite>
			<rules>
				<rule name="Imported Rule 1" stopProcessing="true">
					<match url="^build(.*)$" ignoreCase="false" />
					<action type="Rewrite" url="public/build/{R:1}" />
				</rule>
				<rule name="Imported Rule 2" stopProcessing="true">
					<match url="^bundles(.*)$" ignoreCase="false" />
					<action type="Rewrite" url="public/bundles/{R:1}" />
				</rule>
				<rule name="Imported Rule 3" stopProcessing="true">
					<match url="^robots.txt$" ignoreCase="false" />
					<action type="Rewrite" url="public/robots.txt" />
				</rule>
				<rule name="Imported Rule 4" stopProcessing="true">
					<match url="^(.*)$" ignoreCase="false" />
					<conditions logicalGrouping="MatchAll">
						<add input="{REQUEST_FILENAME}" matchType="IsFile" ignoreCase="false" negate="true" />
						<add input="{REQUEST_FILENAME}" matchType="IsDirectory" ignoreCase="false" negate="true" />
					</conditions>
					<action type="Rewrite" url="public/index.php" appendQueryString="true" />
				</rule>
			</rules>
		</rewrite>
	</system.webServer>
</configuration>
```

I saw that in the same discussion (#195), the user (nadorator) reported the same issue with the design file not being stored correctly and provided a solution there. But I completely missed that part, and I'm sorry since it created trouble for you, too. I will have a look at how to fix that in the next release.

Kind regards,

zepich

**Reply by sukapyx** [reply]
Thank you for the help! I'm glad that we are getting this thing solved.

Currently:

web.config in the public folder:
`<?xml version="1.0" encoding="UTF-8"?>
<configuration>
    <system.webServer>
        <directoryBrowse enabled="false" />
    </system.webServer>
</configuration>
`

web.config in the parent (C:\MyTestWebsite) folder:
`<?xml version="1.0" encoding="UTF-8"?>
<configuration>
    <system.webServer>
        <defaultDocument>
            <files>
                <add value="index.php" />
            </files>
        </defaultDocument>
    </system.webServer>
</configuration>
`

**Reply by zepich** [reply]
Hi @sukapyx 

Can you add the `<rewrite>...</rewrite>` part in your config in the root directory (not in the `/public` directory) as in the config above?

Kind rgards,

zepich

**Reply by sukapyx** [reply]
`<?xml version="1.0" encoding="UTF-8"?>
<configuration>
    <system.webServer>
        <defaultDocument>
            <files>
                <add value="index.php" />
            </files>
        </defaultDocument>
        <rewrite>
			<rules>
				<rule name="Imported Rule 1" stopProcessing="true">
					<match url="^build(.*)$" ignoreCase="false" />
					<action type="Rewrite" url="public/build/{R:1}" />
				</rule>
				<rule name="Imported Rule 2" stopProcessing="true">
					<match url="^bundles(.*)$" ignoreCase="false" />
					<action type="Rewrite" url="public/bundles/{R:1}" />
				</rule>
				<rule name="Imported Rule 3" stopProcessing="true">
					<match url="^robots.txt$" ignoreCase="false" />
					<action type="Rewrite" url="public/robots.txt" />
				</rule>
				<rule name="Imported Rule 4" stopProcessing="true">
					<match url="^(.*)$" ignoreCase="false" />
					<conditions logicalGrouping="MatchAll">
						<add input="{REQUEST_FILENAME}" matchType="IsFile" ignoreCase="false" negate="true" />
						<add input="{REQUEST_FILENAME}" matchType="IsDirectory" ignoreCase="false" negate="true" />
					</conditions>
					<action type="Rewrite" url="public/index.php" appendQueryString="true" />
				</rule>
			</rules>
		</rewrite>
    </system.webServer>
</configuration>`

After stopping and starting the IIS server, no change. Still 
![image](https://github.com/mosparo/mosparo/assets/143611950/666a14db-ccc4-4b33-a781-414564296c72)

**Comment by zepich** [comment]
Hi @sukapyx 

What happens if you add an additional rewrite rule?

```xml
                                 <rule name="Resources Rule" stopProcessing="true">
					<match url="^resources/(.+)/(.+)$" ignoreCase="false" />
					<action type="Rewrite" url="public/resources/{R:1}/{R:2}" />
				</rule>
```

Kind regards,

zepich

**Reply by sukapyx** [reply]
Does not seem to change anything.

**Reply by zepich** [reply]
Hi @sukapyx 

Sorry, it took a while to set up Windows and IIS so I can test it myself.

I did not understand something correctly, so I gave you the wrong information before.

Since you put the site's DocumentRoot in the `public/` directory, the Rewrite rules are in the wrong location and are incorrect.

Please edit the `web.config` in the `public/` directory and insert the following code:

```config
<?xml version="1.0" encoding="UTF-8"?>
<configuration>
    <system.webServer>
        <rewrite>
            <rules>
                <rule name="Imported Rule 1" stopProcessing="true">
                    <match url="^build(.*)$" ignoreCase="false" />
                    <action type="Rewrite" url="build/{R:1}" />
                </rule>
                <rule name="Imported Rule 2" stopProcessing="true">
                    <match url="^bundles(.*)$" ignoreCase="false" />
                    <action type="Rewrite" url="bundles/{R:1}" />
                </rule>
                <rule name="Imported Rule 3" stopProcessing="true">
                    <match url="^robots.txt$" ignoreCase="false" />
                    <action type="Rewrite" url="robots.txt" />
                </rule>
                <rule name="Imported Rule 4" stopProcessing="true">
                    <match url="^(.*)$" ignoreCase="false" />
                    <conditions logicalGrouping="MatchAll">
                        <add input="{REQUEST_FILENAME}" matchType="IsFile" negate="true" />
                        <add input="{REQUEST_FILENAME}" matchType="IsDirectory" negate="true" />
                    </conditions>
                    <action type="Rewrite" url="index.php" appendQueryString="true" />
                </rule>
            </rules>
        </rewrite>
        <defaultDocument>
            <files>
                <add value="index.php" />
            </files>
        </defaultDocument>
    </system.webServer>
</configuration>
```

(**Edit:** After I wrote this post, I realized that the first three rules are not even needed. It's not a problem but in theory, they're not needed I guess)

With this configuration, you should be able to use mosparo without the `index.php` in the URL (just https://localhost/login to open mosparo). All URLs should be correctly rewritten.

If you open the following URL, you should get only one redirect to the correct CSS file:

`https://localhost/resources/<uuid>.css`

(Please replace `<uuid>` with the UUID you see in the project settings. I was too lazy to type it from your screenshots ;))

Can you confirm that this redirect works correctly now with the rewrite configuration I posted above?

If so, please adjust one more thing. In the demo site's initialization function, the second parameter is the host of your installation. You have `index.php` there right now, but it should be `https://localhost`.

If you adjust these two things, the CSS loading should work correctly.

Please let me know if that works.

Kind regards,

zepich

**Reply by sukapyx** [reply]
Thank you, but unfortunately this does not work.

I left the web.config in the MyTestWebsite folder the same as it was before:
![image](https://github.com/mosparo/mosparo/assets/143611950/f7958571-6f32-44ca-8774-96e3b63fa18d)
...and made the changes you proposed in the MyTestWebsite/public/web.config:
![image](https://github.com/mosparo/mosparo/assets/143611950/cb70aac0-aee5-4a68-baeb-0493996c431e)
but this produces:
![image](https://github.com/mosparo/mosparo/assets/143611950/1756de32-3169-4704-8bc3-8c81e8aa4353)

IIS configuration:
![image](https://github.com/mosparo/mosparo/assets/143611950/73bb1f0a-ba37-4967-9166-6ae946080494)

...clicking any of the icons. e.g. Authentication, CGI etc now produces error:
![image](https://github.com/mosparo/mosparo/assets/143611950/059c582b-1ecc-42c2-a384-741b496b07af)

Rights to the web.config should be set as needed...
![image](https://github.com/mosparo/mosparo/assets/143611950/646c67d0-6401-44ab-9b73-bf0a8b063171)

**Reply by zepich** [reply]
Do you have the IIS rewrite Module installed and enabled?

**Reply by sukapyx** [reply]
I installed the IIS rewrite Module and everything works now! Thank you sir! Appreciation.

![image](https://github.com/mosparo/mosparo/assets/143611950/cff8453d-9638-4d27-82a6-d03b149aafc6)

**Reply by zepich** [reply]
Awesome! Thank you very much for the feedback!

**Comment by zepich** [comment]
Hi @sukapyx 

So, regarding your other questions.

### Hosts

The hosts in the project settings are the domains on which you use a project. So, if you host your website under the address `www.example.com`, you have to specify that domain (`www.example.com`) as the host in the project settings.

If your website is hosted under `www.example.com` and `example.com`, you must specify both domains. If you use the same project for another website, like `test-website.example.com`, you must also specify this domain as a host in the project settings.

But it's always the domain, not the protocol at the beginning, and not the path after the domain. So, in your case, it would be simply `localhost`.

### Update

I will test the update process on my IIS over the weekend and let you know what you can do to update mosparo correctly.

Do you have any other questions, or did I miss something?

Kind regards,

zepich

**Reply by sukapyx** [reply]
Thanks! I think if we get the update issue resolved as well then it's perfect.

**Reply by zepich** [reply]
Hi @sukapyx 

I've investigated the issue with the update. It's a combination of multiple problems.

In general, the update process takes too long and hits the `max_execution_time` limit. But even with increasing that, there are other issues, mainly regarding file paths (`\` instead of `/`, case-insensitive paths). I'll prepare fixes for these issues.

Additionally, I started investigating the slow performance and found at least one performance issue that we can optimize for all systems to speed up the update process.

It will take a moment until all issues are resolved. After that, the update system should work correctly on Windows, too.

Kind regards,

zepich

**Reply by sukapyx** [reply]
Hello Zepich!

Thank you for the update!

BR

**Reply by zepich** [reply]
Hi @sukapyx 

Thank you very much for your patience.

The new version of mosparo, v1.2.2, contains a lot of fixes for compatibility with Windows servers. The update process is now fully working. Also, the design CSS file is stored correctly now.

**Important:**  You cannot update from the version you're on right now to this new version. The update system works only from version 1.2.2 onwards. To update your installation now, you have to download the ZIP file manually, delete the existing files, and copy the new files into the directory (as for a new installation). Please backup the configuration file in `/config/env.mosparo.php` before you delete the files and copy it back after you copy the new files from mosparo.

Also important is that the update process is slow on Windows. I will continue working on that part, but for now, you should at least set the timeout in PHP and IIS to 60 seconds or more. In my tests, it usually took 40 - 50 seconds on a freshly started server.

Alternatively, you can use the CLI to update mosparo in the future. There, you do not have a timeout.

Thank you very much for your help!

Kind regards,

zepich
